flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterSchemaConverter.java

```diff
@@ -417,6 +422,92 @@ void checkColumnExists(String columnName) {
+                FlinkTypeFactory typeFactory,
+                SqlValidator sqlValidator,
+                Consumer<SqlTableConstraint> constraintValidator,
+                Function<SqlNode, String> escapeExpressions,
+                SchemaResolver schemaResolver) {
+            super(
+                    originalSchema,
+                    typeFactory,
+                    sqlValidator,
+                    constraintValidator,
+                    escapeExpressions,
+                    schemaResolver);
+        }
+
+        @Override
+        void checkColumnExists(String columnName) {
+            if (!sortedColumnNames.contains(columnName)) {
+                throw new ValidationException(
+                        String.format(
+                                "%sTry to modify a column `%s` which does not exist in the table.",
+                                EX_MSG_PREFIX, columnName));
+            }
+        }
+
+        @Override
+        void checkPrimaryKeyExists() {
+            if (primaryKey == null) {
+                throw new ValidationException(
+                        String.format(
+                                "%sThe base table does not define any primary key constraint. You might "
+                                        + "want to add a new one.",
+                                EX_MSG_PREFIX));
+            }
+        }
+
+        @Override
+        void checkWatermarkExists() {
+            if (watermarkSpec == null) {
+                throw new ValidationException(
+                        String.format(
+                                "%sThe base table does not define any watermark. You might "
+                                        + "want to add a new one.",
+                                EX_MSG_PREFIX));
+            }
+        }
+
+        @Override
+        Optional<Integer> getColumnPosition(SqlTableColumnPosition columnPosition) {
+            if (columnPosition.isFirstColumn() || columnPosition.isAfterReferencedColumn()) {
+                sortedColumnNames.remove(columnPosition.getColumn().getName().getSimple());
+                return super.getColumnPosition(columnPosition);
+            }
+            return Optional.empty();
+        }
+
+        @Override
+        Schema.UnresolvedPhysicalColumn convertPhysicalColumn(
+                SqlTableColumn.SqlRegularColumn physicalColumn) {
+            Schema.UnresolvedPhysicalColumn newColumn = super.convertPhysicalColumn(physicalColumn);
+            String columnName = newColumn.getName();
+            // preserves the primary key's nullability
+            if (primaryKey != null && primaryKey.getColumnNames().contains(columnName)) {
+                newColumn =
+                        new Schema.UnresolvedPhysicalColumn(
+                                columnName,
+                                newColumn.getDataType().notNull(),
+                                newColumn.getComment().orElse(null));
+            }
```

Comment:
Why not mark the column is not null when building the final schema with the final pk?

Reply:
> Why not mark the column is not null when building the final schema with the final pk?

Because under some conditions the final pk does not get a chance to be updated.
Consider the following case
```sql
create table T (
  f0 int,
  f1 string primary key not enforced, -- f1 is converted to string not null implicitly
  f2 double
) with (
  ...
);

-- then change f1 pos
alter table T modify f1 string first
```

For this case, L#103`alterTableSchema.getFullConstraint().ifPresent(converter::updatePrimaryKey);` does not get executed.

Reply:
I mean in the `SchemaConverter#convert`. BTW, do you think we need to modify the type to not null here. As far as I know, in the `SqlAlterTableSchema#validate` we have already mark used column not null.

Reply:
> in the SqlAlterTableSchema#validate we have already mark used column not null.

Actually, `SqlAlterTableSchema#validate` is not suitable under this condition. Consider the previous case I mentioned, 
`SqlAlterTableSchema#validate` will do nothing because this change does not reveal any primary key info.